builtin/providers/aws/resource_aws_lambda_permission.go

```diff
@@ -189,7 +189,28 @@ func resourceAwsLambdaPermissionRead(d *schema.ResourceData, meta interface{}) e
 		statement, err = findLambdaPolicyStatementById(&policy, d.Id())
 		return resource.RetryableError(err)
 	})
+
 	if err != nil {
+		switch e := err.(type) {
+		case *resource.TimeoutError:
+			switch underlying := e.LastError.(type) {
+			case awserr.Error:
+				if underlying.Code() == "ResourceNotFoundException" {
+					log.Printf("[WARN] No Lambda Permission Policy found: %v", input)
+					d.SetId("")
+					return nil
+				}
+			case *resource.NotFoundError:
+				log.Printf("[WARN] found Lambda Permission Policy for %v but statement '%s' is absent", input, d.Id())
+				d.SetId("")
+				return nil
+			}
+		case (*resource.NotFoundError):
```

Comment:
> NotFoundChecks gets defaulted to 20, so if the check ran > 20 times in 1 minute, then NotFoundError could be returned.

You're right, I must have been in some kind of haze to not see that on Friday. Sorry.

I just tried to simulate the situation by following these steps:

1. `terraform apply` (containing `aws_lambda_permission`)
2. `aws lambda remove-permission --function-name=lambda_function_name --statement-id=TheStatementId`
3. `terraform plan`

which made me think that `e := err.(type)` in fact gets detected as `*awserr.requestError` - hence completely bypasses the whole `switch` which in turn leads me to 2 thoughts:

1. The golang type switch doesn't reflect inheritance
2. The `TimeoutError` is never returned from the `resource.Retry` wrapper, errors coming from `WaitForState` function directly [are ignored](https://github.com/hashicorp/terraform/blob/master/helper/resource/wait.go#L39) and any last error which came from `RefreshFunc` is returned instead. Based on https://github.com/hashicorp/terraform/commit/5160578e186230aecb251bdcfb0219cac35c9230 I assume this was an intention.